### ISSUE TITLE
Add graceful shutdown to systemd unit

### ISSUE THREAD

**Issue by bharrisau** [lichess-org/fishnet#260]
It's a bit hacky to get around the lack of asynchronous stop commands. But I've got something like this to give a 5 minute window to the process to finish any remaining work.

```
[Service]
ExecStart=/home/fishnet/fishnet --auto-update --conf /home/fishnet/fishnet.ini run
ExecStop=/bin/kill -INT "$MAINPID"
ExecStop=/usr/bin/sh -c 'while kill -0 "$MAINPID" 2>/dev/null; do sleep 1; done'
KillMode=mixed
TimeoutStopSec=300
```

**Comment by niklasf** [comment]
Generally such long timeouts for shutdown are not acceptable, so instead the default `KillSignal=SIGTERM` is used. This will cause a somewhat graceful shutdown, i.e., work won't be completed, but the server is notified.

The hack above can be simplified to:

```
KillMode=mixed
KillSignal=SIGINT
TimeoutStopSec=300
```

**Comment by bharrisau** [comment]
You don't get the SIGTERM then though, just straight to SIGKILL if the batch takes too long. That might be worse as the server isn't notified or anything.

Yes, the 5 minute time isn't for everyone. Just shared it incase it was handy.

On Thu, 22 Feb 2024, at 2:42 AM, Niklas Fiekas wrote:
> 
> 
> Generally such long timeouts for shutdown are not acceptable, so instead the default `KillSignal=SIGTERM` is used. This will cause a somewhat graceful shutdown, i.e., work won't be completed, but the server is notified.
> 
> The hack above can be simplified to:
> 
> `KillMode=mixed
> KillSignal=SIGINT
> TimeoutStopSec=300
`

**Comment by niklasf** [comment]
Ah, interesting. I didn't realize how nicely your snippet works: So it's immediate SIGINT, *SIGTERM* after the timeout, and then eventually SIGKILL if something went wrong?

**Comment by bharrisau** [comment]
Yeah. But there is only 1 timeout value, so it's the same between SIGINT to SIGTERM, and SIGTERM to SIGKILL.

**Comment by vpag** [comment]
I've just spent some time going down the rabbit hole of setting up a truly graceful shutdown for a rootless quadlet (podman+systemd) with fishnet on board, and I've come to a rather annoying conclusion: there's no two-unit-files-fit-all solution.

`KillSignal=SIGINT` works as expected (kudos++ for #74), but only for manual `systemctl restart/stop`, not for a system shutdown (which is what I was aiming for).

----

Just sharing my findings for the **userland** + **shutdown** case:

You can set `StopTimeout` (in the Container section of a quadlet) or `TimeoutStopSec` (in the Service section of a systemd unit) to whatever you want, but for rootless containers and user-level units, these values are effectively ignored during a system shutdown/reboot/sleep.

The culprit is `systemd-logind`. By default, it only allows a 30-second window for all user processes to terminate before the system-level manager hard-kills the entire user session (regardless of whether _lingering_ is enabled, you see).

Users could be advised to increase `InhibitDelayMaxSec` and potentially use a system-wide inhibitor lock (via `systemd-inhibit ...` as an `ExecStop` action) to prevent the OS from unmounting filesystems (and cutting off `.fishnet-stats` dumping) and going offline, before the program has completed its batches.

Without these _host-side tweaks_, and taking care of dependencies ({Before, After, Requires} x {network, mounts}) a graceful shutdown for (at least) rootless containers during a reboot is mostly a myth.

Personally, I finally settled on a workaround with a `/usr/local/sbin/graceful-shutdown` wrapper like

```sh
/usr/bin/podman kill -s SIGINT systemd-fishnet
# ... (+some logic for my other services)
/usr/sbin/shutdown "$@"
```
HTH.
